File: ios/media_filters/Sources/media_filters/VideoPreviewManager.swift
import Foundation

public class VideoPreviewManager {
    nonisolated(unsafe) public static let instance = VideoPreviewManager()

    private let lock = NSLock()
    private var previews: [Int: VideoPreview] = [:]

    private init() {}
    
    func getCount() -> Int {
        return previews.count;
    }

    func createPreview(viewId: Int) -> VideoPreview {
        lock.lock()
        defer { lock.unlock() }
        
        if let existingPreview = previews[viewId] {
            return existingPreview
        }

        let preview = VideoPreview(id: viewId)
        previews[viewId] = preview

        return preview
    }

    func getPreview(_ viewId: Int) -> VideoPreview? {
        lock.lock()
        defer { lock.unlock() }
        return previews[viewId]
    }

    func destroyPreview(viewId: Int) {
        lock.lock()
        defer { lock.unlock() }

        if let preview = previews[viewId] {
            previews.removeValue(forKey: viewId)
            
            Task { @MainActor in
                preview.cleanup()
            }
        }
    }

    @MainActor func destroyAllPreviews() {
        lock.lock()
        defer { lock.unlock() }

        for preview in previews.values {
            preview.cleanup()
        }
        previews.removeAll()
    }
}

File: android/src/main/kotlin/dev/arish/media_filters/VideoPreviewManager.kt
package dev.arish.media_filters

import android.content.Context
import android.util.Log
import java.util.concurrent.ConcurrentHashMap

class VideoPreviewManager private constructor() {
    private val previews = ConcurrentHashMap<Int, VideoPreview>()
    
    companion object {
        @Volatile
        private var INSTANCE: VideoPreviewManager? = null
        
        fun getInstance(): VideoPreviewManager {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: VideoPreviewManager().also { INSTANCE = it }
            }
        }
        
        private const val TAG = "VideoPreviewManager"
    }
    
    fun getCount(): Int = previews.size
    
    fun createPreview(viewId: Int, context: Context): VideoPreview {
        return previews.getOrPut(viewId) {
            Log.d(TAG, "Creating new video preview with ID: $viewId")
            VideoPreview(viewId, context)
        }
    }
    
    fun getPreview(viewId: Int): VideoPreview? {
        return previews[viewId]
    }
    
    fun destroyPreview(viewId: Int) {
        previews.remove(viewId)?.let { preview ->
            Log.d(TAG, "Destroying video preview with ID: $viewId")
            preview.cleanup()
        }
    }
    
    fun destroyAllPreviews() {
        Log.d(TAG, "Destroying all video previews")
        previews.values.forEach { it.cleanup() }
        previews.clear()
    }
}